File: backend/app/services/task_runner.py
"""
Agent task poller: uses LISTEN/NOTIFY to wake on new tasks,
with polling fallback for stale task recovery.

Sidecar bridge:
  Frontend creates task row -> NOTIFY wakes poller -> LangGraph executes -> DB updated -> UI sees result
"""

import asyncio
import json
import logging
import traceback

from app.config import settings
from app.db import get_pool
from graph.workflows.summarize import run_summarize_workflow
from graph.workflows.triage import run_triage_workflow
from graph.workflows.agent import run_agent_workflow

logger = logging.getLogger(__name__)

WORKFLOW_HANDLERS = {
    "summarize": run_summarize_workflow,
    "triage": run_triage_workflow,
    "agent": run_agent_workflow,
    "custom": run_agent_workflow,  # custom uses configurable agent definitions
}

STALE_TASK_TIMEOUT_MINUTES = 10


async def _reclaim_stale_tasks() -> None:
    """Reset tasks stuck in 'running' with no recent heartbeat back to 'failed'."""
    try:
        pool = get_pool()
        rows = await pool.fetch("""
            UPDATE agent_tasks
            SET status = 'failed',
                error = 'Worker timeout: task exceeded heartbeat deadline',
                completed_at = now(),
                updated_at = now()
            WHERE status = 'running'
              AND heartbeat_at < now() - interval '%s minutes'
            RETURNING id
        """ % STALE_TASK_TIMEOUT_MINUTES)
        if rows:
            logger.warning("Reclaimed %d stale running tasks", len(rows))
    except Exception as e:
        logger.error("Stale task reclaim error: %s", e)


async def _claim_and_run_one() -> bool:
    """Try to claim and execute one pending task. Returns True if a task was processed."""
    pool = get_pool()

    row = await pool.fetchrow("""
        UPDATE agent_tasks
        SET status = 'running',
            started_at = now(),
            heartbeat_at = now(),
            updated_at = now()
        WHERE id = (
            SELECT id FROM agent_tasks
            WHERE status = 'pending'
            ORDER BY created_at ASC
            LIMIT 1
            FOR UPDATE SKIP LOCKED
        )
        RETURNING id, workflow, input, node_id, owner_id, conversation_id, agent_definition_id
    """)

    if row is None:
        return False

    task_id = row["id"]
    workflow = row["workflow"]
    task_input = dict(row["input"]) if row["input"] else {}
    node_id = row["node_id"]
    owner_id = row["owner_id"]
    conversation_id = row["conversation_id"]
    agent_def_id = row["agent_definition_id"]

    # Inject conversation_id and agent_definition_id into task input
    if conversation_id:
        task_input["conversation_id"] = str(conversation_id)
    if agent_def_id:
        task_input["agent_definition_id"] = str(agent_def_id)

    logger.info("Running task %s (workflow=%s)", task_id, workflow)

    handler = WORKFLOW_HANDLERS.get(workflow)
    if handler is None:
        await pool.execute("""
            UPDATE agent_tasks
            SET status = 'failed', error = $2, completed_at = now(), updated_at = now()
            WHERE id = $1
        """, task_id, f"Unknown workflow: {workflow}")
        return True

    try:
        async def heartbeat_loop() -> None:
            while True:
                await asyncio.sleep(30)
                await pool.execute(
                    "UPDATE agent_tasks SET heartbeat_at = now() WHERE id = $1",
                    task_id,
                )

        heartbeat_task = asyncio.create_task(heartbeat_loop())
        try:
            result = await handler(
                task_id=str(task_id),
                node_id=str(node_id) if node_id else None,
                owner_id=str(owner_id),
                task_input=task_input,
            )
        finally:
            heartbeat_task.cancel()
            try:
                await heartbeat_task
            except asyncio.CancelledError:
                pass

        if result and result.get("error"):
            await pool.execute("""
                UPDATE agent_tasks
                SET status = 'failed', error = $2, completed_at = now(), updated_at = now()
                WHERE id = $1
            """, task_id, result["error"])
            logger.warning("Task %s failed: %s", task_id, result["error"])
            return True

        await pool.execute("""
            UPDATE agent_tasks
            SET status = 'completed',
                output = $2::jsonb,
                completed_at = now(),
                updated_at = now()
            WHERE id = $1
        """, task_id, json.dumps(result or {}))

        logger.info("Task %s completed", task_id)

    except Exception as e:
        tb = traceback.format_exc()
        logger.error("Task %s failed: %s", task_id, str(e))
        await pool.execute("""
            UPDATE agent_tasks
            SET status = 'failed',
                error = $2,
                completed_at = now(),
                updated_at = now()
            WHERE id = $1
        """, task_id, f"{type(e).__name__}: {str(e)}\n{tb[:2000]}")

    return True


async def poll_agent_tasks() -> None:
    """
    Main loop: LISTEN for new_task notifications, with polling fallback.
    Also periodically reclaims stale running tasks.
    """
    pool = get_pool()
    logger.info("Task poller started (fallback interval=%ds)", settings.task_poll_interval_seconds)

    notify_event = asyncio.Event()

    async def _listen_loop() -> None:
        conn = await pool.acquire()
        try:
            await conn.add_listener("new_task", lambda *_args: notify_event.set())
            while True:
                await asyncio.sleep(3600)
        finally:
            await conn.remove_listener("new_task", lambda *_args: None)
            await pool.release(conn)

    listener_task = asyncio.create_task(_listen_loop())
    stale_check_counter = 0

    try:
        while True:
            try:
                await asyncio.wait_for(
                    notify_event.wait(),
                    timeout=settings.task_poll_interval_seconds,
                )
            except asyncio.TimeoutError:
                pass
            notify_event.clear()

            while await _claim_and_run_one():
                pass

            stale_check_counter += 1
            if stale_check_counter >= 12:
                stale_check_counter = 0
                await _reclaim_stale_tasks()

    except asyncio.CancelledError:
        logger.info("Task poller shutting down")
        listener_task.cancel()
        try:
            await listener_task
        except asyncio.CancelledError:
            pass
        raise

File: frontend/worker/task-runner.ts
/**
 * Task runner: claims pending agent_tasks, runs workflows, handles heartbeat
 * and stale task reclamation. Mirrors backend/app/services/task_runner.py.
 */

import { sql } from "drizzle-orm";
import { getDb } from "./db";
import { logger } from "./logger";

const HEARTBEAT_INTERVAL_MS = 30_000;
const STALE_TASK_TIMEOUT_MINUTES = 10;
const STALE_CHECK_INTERVAL_MS = 60_000;

export type WorkflowHandler = (
  taskId: string,
  nodeId: string | null,
  ownerId: string,
  input: Record<string, unknown>,
) => Promise<Record<string, unknown>>;

interface ClaimedTask {
  id: string;
  workflow: string;
  input: Record<string, unknown>;
  nodeId: string | null;
  ownerId: string;
  conversationId: string | null;
  memberId: string | null;
}

let handlers: Record<string, WorkflowHandler> = {};
let allowedWorkflows: Set<string> | null = null;

export function registerHandler(workflow: string, handler: WorkflowHandler) {
  handlers[workflow] = handler;
}

export function setAllowedWorkflows(workflows: string[] | null) {
  allowedWorkflows = workflows ? new Set(workflows) : null;
}

// ---------------------------------------------------------------------------
// Claim
// ---------------------------------------------------------------------------

async function claimOne(): Promise<ClaimedTask | null> {
  const db = getDb();

  const workflowFilter = allowedWorkflows
    ? sql`AND workflow = ANY(${Array.from(allowedWorkflows)}::workflow_type[])`
    : sql``;

  const rows = await db.execute(sql`
    UPDATE agent_tasks
    SET status = 'running',
        started_at = now(),
        heartbeat_at = now(),
        updated_at = now()
    WHERE id = (
      SELECT id FROM agent_tasks
      WHERE status = 'pending' ${workflowFilter}
      ORDER BY created_at ASC
      LIMIT 1
      FOR UPDATE SKIP LOCKED
    )
    RETURNING id, workflow, input, node_id AS "nodeId",
              owner_id AS "ownerId", conversation_id AS "conversationId",
              member_id AS "memberId"
  `);

  if (!rows[0]) return null;
  const r = rows[0] as unknown as ClaimedTask;
  return {
    id: r.id,
    workflow: r.workflow,
    input: (typeof r.input === "string" ? JSON.parse(r.input) : r.input) as Record<string, unknown>,
    nodeId: r.nodeId,
    ownerId: r.ownerId,
    conversationId: r.conversationId,
    memberId: r.memberId,
  };
}

// ---------------------------------------------------------------------------
// Heartbeat
// ---------------------------------------------------------------------------

function startHeartbeat(taskId: string): NodeJS.Timeout {
  const db = getDb();
  return setInterval(async () => {
    try {
      await db.execute(sql`
        UPDATE agent_tasks SET heartbeat_at = now() WHERE id = ${taskId}::uuid
      `);
    } catch (e) {
      logger.warn(`Heartbeat failed for task ${taskId}:`, e);
    }
  }, HEARTBEAT_INTERVAL_MS);
}

// ---------------------------------------------------------------------------
// Complete / Fail
// ---------------------------------------------------------------------------

async function completeTask(taskId: string, output: Record<string, unknown>) {
  const db = getDb();
  await db.execute(sql`
    UPDATE agent_tasks
    SET status = 'completed',
        output = ${JSON.stringify(output)}::jsonb,
        completed_at = now(),
        updated_at = now()
    WHERE id = ${taskId}::uuid
  `);
}

async function failTask(taskId: string, error: string) {
  const db = getDb();
  await db.execute(sql`
    UPDATE agent_tasks
    SET status = 'failed',
        error = ${error},
        completed_at = now(),
        updated_at = now()
    WHERE id = ${taskId}::uuid
  `);
}

// ---------------------------------------------------------------------------
// Stale reclaim
// ---------------------------------------------------------------------------

export async function reclaimStaleTasks() {
  const db = getDb();
  const rows = await db.execute(sql`
    UPDATE agent_tasks
    SET status = 'failed',
        error = 'Worker timeout: task exceeded heartbeat deadline',
        completed_at = now(),
        updated_at = now()
    WHERE status = 'running'
      AND heartbeat_at < now() - make_interval(mins => ${STALE_TASK_TIMEOUT_MINUTES})
    RETURNING id
  `);
  const count = (rows as unknown[]).length;
  if (count > 0) {
    logger.info(`Reclaimed ${count} stale task(s)`);
  }
}

// ---------------------------------------------------------------------------
// Run one task
// ---------------------------------------------------------------------------

async function runTask(task: ClaimedTask) {
  const handler = handlers[task.workflow];
  if (!handler) {
    await failTask(task.id, `Unknown workflow: ${task.workflow}`);
    return;
  }

  const hb = startHeartbeat(task.id);
  try {
    // Inject conversation_id / member_id / channel_id into input
    const input = { ...task.input };
    if (task.conversationId) input.conversation_id = task.conversationId;
    if (task.memberId) input.member_id = task.memberId;
    if (task.nodeId) input.node_id = task.nodeId;

    const output = await handler(task.id, task.nodeId, task.ownerId, input);

    if (output.error) {
      await failTask(task.id, String(output.error));
    } else {
      await completeTask(task.id, output);
    }
  } catch (e: unknown) {
    const msg = e instanceof Error ? e.message : String(e);
    logger.error(`Task ${task.id} (${task.workflow}) failed:`, msg);
    await failTask(task.id, msg);
  } finally {
    clearInterval(hb);
  }
}

// ---------------------------------------------------------------------------
// Poll loop
// ---------------------------------------------------------------------------

let running = true;

export function stopTaskRunner() {
  running = false;
}

export interface WakeRef {
  current: Promise<void>;
}

/**
 * Main poll loop: wait for LISTEN notification or fallback timeout,
 * then drain all pending tasks. WakeRef allows the LISTEN callback to
 * resolve the current promise and create the next one for future wakes.
 */
export async function pollLoop(wakeRef: WakeRef) {
  while (running) {
    await Promise.race([
      wakeRef.current,
      new Promise<void>((r) => setTimeout(r, 5_000)),
    ]);

    let claimed = true;
    while (claimed && running) {
      const task = await claimOne();
      if (task) {
        logger.info(`Running task ${task.id} [${task.workflow}]`);
        await runTask(task);
      } else {
        claimed = false;
      }
    }
  }
}

/**
 * Periodic stale-task reclaim loop.
 */
export async function staleReclaimLoop() {
  while (running) {
    await new Promise<void>((r) => setTimeout(r, STALE_CHECK_INTERVAL_MS));
    if (!running) break;
    try {
      await reclaimStaleTasks();
    } catch (e) {
      logger.warn("Stale reclaim error:", e);
    }
  }
}
